In [57]:
import pandas as pd
import numpy as np

import matplotlib
import matplotlib.pyplot as plt
import seaborn as sns

from scipy.stats import chi2_contingency
%matplotlib inline

import warnings
warnings.filterwarnings('ignore')

df = pd.read_csv("/home/piai/preprocessed_scale.csv")
df.head()

scale steel_kind  pt_thick  pt_width  pt_length hsb  fur_heat_temp  \
0      0          T        32      3700      15100  적용           1144   
1      0          T        32      3700      15100  적용           1144   
2      0          T        33      3600      19200  적용           1129   
3      0          T        33      3600      19200  적용           1152   
4      0          T        38      3100      13300  적용           1140   

   fur_heat_time  fur_soak_temp  fur_soak_time rolling_method  rolling_temp  \
0            116           1133             59     TMCP(온도제어)           934   
1            122           1135             53     TMCP(온도제어)           937   
2            116           1121             55     TMCP(온도제어)           889   
3            125           1127             68     TMCP(온도제어)           885   
4            134           1128             48     TMCP(온도제어)           873   

   descaling_count  fur_preheat_time furnace  
0                8                84   1호기1열  
1                8                63   1호기2열  
2                8                87   2호기1열  
3                8                73   2호기2열  
4                8                64   3호기1열

카이제곱 검정

scale과 steel_kind 카이제곱 검정 

In [27]:
df_scale = df['scale']
df_scale

0      0
1      0
2      0
3      0
4      0
      ..
989    0
990    0
991    0
992    0
993    0
Name: scale, Length: 994, dtype: int64

In [11]:
df_steel_kind = df['steel_kind']
df_steel_kind

0      T
1      T
2      T
3      T
4      T
      ..
995    C
996    C
997    C
998    C
999    C
Name: steel_kind, Length: 1000, dtype: object

In [34]:
contingency_table = pd.crosstab(df_scale, df_steel_kind)

chi2, p, dof, expected = chi2_contingency(contingency_table)

# 결과 출력
print("Chi-squared Statistic:", chi2)
print("p-value:", p)
print("Degrees of Freedom:", dof)
print("Expected Frequencies:\n", expected)

Chi-squared Statistic: 78.89479838930058
p-value: 6.550700807797746e-19
Degrees of Freedom: 1
Expected Frequencies:
 [[515.4084507 168.5915493]
 [233.5915493  76.4084507]]


p-value가 6.55×10−19로, 일반적인 유의 수준(예: 0.05)로 0.05보다 훨씬 작음. 따라서,  df_scale과 df_steel_kind 변수 간에 통계적으로 
유의미한 관계가 있음을 의미

scale과 steel_hsb 카이제곱 검정 

In [40]:
df_steel_hsb = df['hsb']
df_steel_hsb

0      적용
1      적용
2      적용
3      적용
4      적용
       ..
989    적용
990    적용
991    적용
992    적용
993    적용
Name: hsb, Length: 994, dtype: object

In [41]:
value_counts = df['hsb'].value_counts()
print(value_counts)

적용     947
미적용     47
Name: hsb, dtype: int64


In [42]:
contingency_table = pd.crosstab(df_scale, df_steel_hsb)

chi2, p, dof, expected = chi2_contingency(contingency_table)

# 결과 출력
print("Chi-squared Statistic:", chi2)
print("p-value:", p)
print("Degrees of Freedom:", dof)
print("Expected Frequencies:\n", expected)

Chi-squared Statistic: 105.51048606504317
p-value: 9.439705302426995e-25
Degrees of Freedom: 1
Expected Frequencies:
 [[ 32.34205231 651.65794769]
 [ 14.65794769 295.34205231]]


p-value가 9.44e-25로, 유의 수준(일반적으로 0.05)보다 훨씬 작으므로, df_scale과 df_steel_hsb 변수 간에 통계적으로 유의미한 관계가 있다

scale과 furnace 카이제곱 검정 

In [44]:
value_counts = df['furnace'].value_counts()
print(value_counts)

3호기1열    176
1호기2열    174
2호기1열    171
3호기2열    162
1호기1열    158
2호기2열    153
Name: furnace, dtype: int64


In [46]:
df_steel_fur_no = df['furnace']
df_steel_fur_no

0      1호기1열
1      1호기2열
2      2호기1열
3      2호기2열
4      3호기1열
       ...  
989    1호기2열
990    1호기2열
991    2호기2열
992    2호기2열
993    3호기1열
Name: furnace, Length: 994, dtype: object

In [47]:
contingency_table = pd.crosstab(df_scale, df_steel_fur_no)

chi2, p, dof, expected = chi2_contingency(contingency_table)

# 결과 출력
print("Chi-squared Statistic:", chi2)
print("p-value:", p)
print("Degrees of Freedom:", dof)
print("Expected Frequencies:\n", expected)

Chi-squared Statistic: 7.67460098873368
p-value: 0.1751043391036752
Degrees of Freedom: 5
Expected Frequencies:
 [[108.72434608 119.73440644 117.67002012 105.28370221 121.11066398
  111.47686117]
 [ 49.27565392  54.26559356  53.32997988  47.71629779  54.88933602
   50.52313883]]


p-value가 0.175로, 일반적인 유의 수준(예: 0.05)보다 크므로, df_scale과 df_steel_hsb 변수 간에 통계적으로 유의미한 관계가 없다

scale과 rolling_method 카이제곱 검정 

In [48]:
value_counts = df['rolling_method'].value_counts()
print(value_counts)

CR(제어압연)      836
TMCP(온도제어)    158
Name: rolling_method, dtype: int64


In [49]:
df_steel_rolling_method = df['rolling_method']
df_steel_rolling_method 

0      TMCP(온도제어)
1      TMCP(온도제어)
2      TMCP(온도제어)
3      TMCP(온도제어)
4      TMCP(온도제어)
          ...    
989      CR(제어압연)
990      CR(제어압연)
991      CR(제어압연)
992      CR(제어압연)
993      CR(제어압연)
Name: rolling_method, Length: 994, dtype: object

In [50]:
contingency_table = pd.crosstab(df_scale, df_steel_rolling_method)

chi2, p, dof, expected = chi2_contingency(contingency_table)

# 결과 출력
print("Chi-squared Statistic:", chi2)
print("p-value:", p)
print("Degrees of Freedom:", dof)
print("Expected Frequencies:\n", expected)

Chi-squared Statistic: 44.88003167017692
p-value: 2.0948325380804966e-11
Degrees of Freedom: 1
Expected Frequencies:
 [[575.27565392 108.72434608]
 [260.72434608  49.27565392]]


p-value가 2.094e-11로, 일반적인 유의 수준(예: 0.05)보다 훨씬 작으므로, df_scale과 df_steel_hsb 변수 간에 통계적으로 유의미한 관계가 있다

등분산성 검정 

In [56]:
import scipy.stats as stats

# 범주형 변수 목록
categorical_vars = ['pt_thick', 'pt_width', 'fur_heat_temp', 'fur_heat_time', 'pt_length', 'fur_soak_temp', 'fur_soak_time', 'rolling_temp', 'descaling_count','fur_preheat_time']

# 등분산성 검정 결과 저장용 데이터프레임
homogeneity_results = pd.DataFrame(columns=['Category', 'Levene Stat', 'Levene p-value'])

# 등분산성 검정
for i in categorical_vars:
    groups = [group['scale'].values for name, group in df.groupby(i)]
    levene_stat, levene_p = stats.levene(*groups)
    homogeneity_results = homogeneity_results.append({
        'Category': i,
        'Levene Stat': levene_stat,
        'Levene p-value': levene_p
    }, ignore_index=True)

print("\n등분산성 검정 결과:")
print(homogeneity_results)


등분산성 검정 결과:
           Category  Levene Stat  Levene p-value
0          pt_thick     3.489261    2.012906e-13
1          pt_width     3.989145    3.475183e-11
2     fur_heat_temp     1.070799    3.083869e-01
3     fur_heat_time     0.898958    7.455132e-01
4         pt_length     0.743773    9.979552e-01
5     fur_soak_temp     2.379546    9.584696e-09
6     fur_soak_time     1.150656    1.601891e-01
7      rolling_temp     0.957978    6.474176e-01
8   descaling_count    24.877910    1.150331e-23
9  fur_preheat_time     0.647854    9.738107e-01


pt_thick, pt_width, fur_soak_temp, descaling_count는 그룹 간 분산이 통계적으로 유의미하게 다름 따라서, 등분산성을 만족하지 않음 
fur_heat_temp, fur_heat_time, fur_soak_time, rolling_temp, fur_preheat_time는 범주 간 동일한 분산 

SVM 분석

In [58]:
from sklearn.preprocessing import StandardScaler 
from sklearn.model_selection import train_test_split 
from sklearn.svm import SVC
from sklearn.model_selection import GridSearchCV

from sklearn.metrics import accuracy_score, f1_score
from sklearn.metrics import confusion_matrix, classification_report 

df_raw_dummy = pd.get_dummies(df)
df_raw_dummy.head()

scale  pt_thick  pt_width  pt_length  fur_heat_temp  fur_heat_time  \
0      0        32      3700      15100           1144            116   
1      0        32      3700      15100           1144            122   
2      0        33      3600      19200           1129            116   
3      0        33      3600      19200           1152            125   
4      0        38      3100      13300           1140            134   

   fur_soak_temp  fur_soak_time  rolling_temp  descaling_count  ...  hsb_미적용  \
0           1133             59           934                8  ...        0   
1           1135             53           937                8  ...        0   
2           1121             55           889                8  ...        0   
3           1127             68           885                8  ...        0   
4           1128             48           873                8  ...        0   

   hsb_적용  rolling_method_CR(제어압연)  rolling_method_TMCP(온도제어)  furnace_1호기1열  \
0       1                        0                          1              1   
1       1                        0                          1              0   
2       1                        0                          1              0   
3       1                        0                          1              0   
4       1                        0                          1              0   

   furnace_1호기2열  furnace_2호기1열  furnace_2호기2열  furnace_3호기1열  furnace_3호기2열  
0              0              0              0              0              0  
1              1              0              0              0              0  
2              0              1              0              0              0  
3              0              0              1              0              0  
4              0              0              0              1              0  

[5 rows x 23 columns]

In [74]:
# 데이터 분리:설명변수 스케일 변환
df_raw_x = df_raw_dummy.drop("scale", axis = 1, inplace = False)
df_raw_y = df_raw_dummy["scale"] 

v_feature_names = df_raw_x.columns
scaler = StandardScaler() 
df_x_scaled = scaler.fit_transform(df_raw_x)
df_x_scaled = pd.DataFrame(df_x_scaled, columns = v_feature_names)
df_raw_y = df_raw_y.to_frame()
df_x_converted = df_raw_y.join(df_x_scaled)


display(df_x_converted.head())

scale  pt_thick  pt_width  pt_length  fur_heat_temp  fur_heat_time  \
0      0  0.297380  1.758916  -1.568542      -0.629006       1.138138   
1      0  0.297380  1.758916  -1.568542      -0.629006       1.365628   
2      0  0.352984  1.555807  -1.273013      -1.336122       1.138138   
3      0  0.352984  1.555807  -1.273013      -0.251878       1.479374   
4      0  0.631008  0.540262  -1.698287      -0.817571       1.820609   

   fur_soak_temp  fur_soak_time  rolling_temp  descaling_count  ...   hsb_미적용  \
0      -1.042290      -0.617512     -0.098376        -0.353037  ... -0.222779   
1      -0.926564      -0.908186     -0.051371        -0.353037  ... -0.222779   
2      -1.736646      -0.811295     -0.803452        -0.353037  ... -0.222779   
3      -1.389468      -0.181501     -0.866125        -0.353037  ... -0.222779   
4      -1.331605      -1.150415     -1.054145        -0.353037  ... -0.222779   

     hsb_적용  rolling_method_CR(제어압연)  rolling_method_TMCP(온도제어)  \
0  0.222779                -2.300248                   2.300248   
1  0.222779                -2.300248                   2.300248   
2  0.222779                -2.300248                   2.300248   
3  0.222779                -2.300248                   2.300248   
4  0.222779                -2.300248                   2.300248   

   furnace_1호기1열  furnace_1호기2열  furnace_2호기1열  furnace_2호기2열  furnace_3호기1열  \
0       2.300248      -0.460646      -0.455825      -0.426528      -0.463852   
1      -0.434736       2.170862      -0.455825      -0.426528      -0.463852   
2      -0.434736      -0.460646       2.193824      -0.426528      -0.463852   
3      -0.434736      -0.460646      -0.455825       2.344511      -0.463852   
4      -0.434736      -0.460646      -0.455825      -0.426528       2.155859   

   furnace_3호기2열  
0      -0.441261  
1      -0.441261  
2      -0.441261  
3      -0.441261  
4      -0.441261  

[5 rows x 23 columns]

In [83]:
# 데이터 분할 train_test_split(X: 설명변수, Y: 목표변수, test_size = test 데이터 비율)
df_train_x, df_test_x, df_train_y, df_test_y = train_test_split(
    df_x_scaled, df_raw_y, test_size = 0.3, random_state = 1234) 

print("분할 전 설명변수 데이터 :", df_x_scaled.shape)
print("분할 후 설명변수 데이터 :Train", df_train_x.shape, "  Test",df_test_x.shape)

분할 전 설명변수 데이터 : (994, 22)
분할 후 설명변수 데이터 :Train (695, 22)   Test (299, 22)


In [112]:
# 목표변수 빈도 확인

df_train_y
print(df_train_y.value_counts(["scale"]),"\n")
print("scale = 1 비율  ", df_train_y.value_counts(df_train_y["scale"]==1)/len(df_train_y))

scale
0        473
1        222
dtype: int64 

scale = 1 비율   scale
False    0.680576
True     0.319424
dtype: float64


In [113]:
from imblearn.over_sampling import SMOTE
# Over-sampling 설정
sm = SMOTE(sampling_strategy='auto', random_state=1234)


# train데이터를 이용한 Over-sampling
x_resampled, y_resampled = sm.fit_resample(df_train_x,df_train_y)

y_resampled.value_counts()

scale
0        473
1        473
dtype: int64

In [ ]:
para_c = [10**c for c in range(-2, 3)]

para_gamma = [10**gamma for gamma in range(-2, 3)]

estimator = SVC(random_state=1234)

param_grid = {"C":para_c, "gamma":para_gamma}

grid_rf = GridSearchCV(estimator, param_grid, scoring="accuracy", n_jobs = -1)
grid_rf.fit(x_resampled, y_resampled)
print("best estimator model: \n{}".format(grid_rf.best_estimator_))
print("\nbest parameter: \n{}".format(grid_rf.best_params_))
print("\nbest score: \n{}".format(grid_rf.best_score_.round(3)))

train 모델 기준 최적 파라미터로 C=100, gamma=0.1 추천, 정확도는 93.6%

In [117]:
svc_final = SVC(gamma=0.1, C=100, random_state=1234)
svc_final.fit(df_train_x, df_train_y)
y_pred = svc_final.predict(df_test_x)

print("Train Accuracy: {:.3f}".format(svc_final.score(df_train_x, df_train_y)))
print("Test Accuracy: {:.3f}\n".format(svc_final.score(df_test_x, df_test_y)))
print("Test confusion matrix: \n{}".format(confusion_matrix(df_test_y, y_pred)))
print(classification_report(df_test_y, y_pred, digits=3))

Train Accuracy: 1.000
Test Accuracy: 0.873

Test confusion matrix: 
[[192  19]
 [ 19  69]]
              precision    recall  f1-score   support

           0      0.910     0.910     0.910       211
           1      0.784     0.784     0.784        88

    accuracy                          0.873       299
   macro avg      0.847     0.847     0.847       299
weighted avg      0.873     0.873     0.873       299



test 데이터 기준 정분류율은 87.3%, "1"기준 f1 score는 78.4% 

Train Accuracy가 1.000, Test Accuracy가 0.873로 나왔기 때문에 훈련 데이터에서 과도한 성능을 보이는 경향이 있음.
그러나 best score가 0.936로 높은 것을 고려하면, 일부 과적합 경향이 있더라도 모델이 일반화된 성능을 어느 정도 유지하고 있다고 볼 수 있음.

결론:

과적합 가능성은 있지만 심각하지 않을 가능성: Train Accuracy가 100%로 매우 높기 때문에 다소의 과적합은 의심되지만, 교차 검증에서의 best score와 테스트 정확도가 큰 차이를 보이지 않음

일반화 성능이 양호: 현재 설정된 C와 gamma 값으로도 테스트 데이터에서 비교적 안정적인 성능을 유지하고 있으므로, 모델의 일반화 성능은 어느 정도 확보되어 있다고 볼 수 있음.